linkerd/router/src/lib.rs

```diff
@@ -1,577 +1,145 @@
 #![deny(warnings, rust_2018_idioms)]
 
-mod cache;
-pub mod error;
-pub mod layer;
-mod purge;
-
-use self::cache::Cache;
-pub use self::layer::{Config, Layer};
-pub use self::purge::Purge;
-use futures::{Async, Future, Poll};
-use indexmap::IndexMap;
-use linkerd2_lock::Lock;
+use futures::{try_ready, Future, Poll};
+use linkerd2_error::Error;
 use linkerd2_stack::NewService;
-use std::hash::Hash;
-use std::time::Duration;
-pub use tower_load_shed::LoadShed;
-use tracing::{debug, trace};
-
-/// Routes requests based on a configurable `Key`.
-pub struct Router<Req, Rec, N>
-where
-    Rec: Recognize<Req>,
-    N: NewService<Rec::Target>,
-    N::Service: tower::Service<Req>,
-{
-    inner: Inner<Req, Rec, N>,
-    _hangup: purge::Handle,
-}
+use tower::util::{Oneshot, ServiceExt};
+use tracing::trace;
 
-/// Provides a strategy for routing a Request to a Service.
-///
-/// Implementors must provide a `Key` type that identifies each unique route. The
-/// `recognize()` method is used to determine the target for a given request. This target is
-/// used to look up a route in a cache (i.e. in `Router`), or can be passed to
-/// `bind_service` to instantiate the identified route.
-pub trait Recognize<Request> {
-    /// Identifies a Route.
-    type Target: Clone + Eq + Hash;
+pub trait Key<T> {
+    type Key: Clone;
 
-    /// Determines the target for a route to handle the given request.
-    fn recognize(&self, request: &Request) -> Option<Self::Target>;
+    fn key(&self, t: &T) -> Self::Key;
 }
```

Comment:
nitpicky, but: using the name `Key` for the trait that returns a key is not exactly my favorite naming...especially since the associated type is _also_ called `Key`...IMO, the old `Recognize` was more descriptive for the trait. that said...i acknowledge that this may not actually matter.